From 14e767ce50de97969e3e57cb9d26173f0a1aad57 Mon Sep 17 00:00:00 2001
From: Jan Holthuis <jan.holthuis@ruhr-uni-bochum.de>
Date: Wed, 17 Aug 2022 19:47:43 +0200
Subject: [PATCH] Enable support for `none` cipher

---
 cipher.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/cipher.c b/cipher.c
index c3cd5dcf..39b4b8d0 100644
--- a/cipher.c
+++ b/cipher.c
@@ -72,7 +72,7 @@ struct sshcipher {
 #define CFLAG_CHACHAPOLY	(1<<1)
 #define CFLAG_AESCTR		(1<<2)
 #define CFLAG_NONE		(1<<3)
-#define CFLAG_INTERNAL		CFLAG_NONE /* Don't use "none" for packets */
+#define CFLAG_INTERNAL		0
 #ifdef WITH_OPENSSL
 	const EVP_CIPHER	*(*evptype)(void);
 #else
--
2.37.1
