From 9f4c5ec08b586cd5fab13b30ad9139f43202c0a8 Mon Sep 17 00:00:00 2001
From: Jan Holthuis <jan.holthuis@ruhr-uni-bochum.de>
Date: Wed, 17 Aug 2022 19:39:40 +0200
Subject: [PATCH] Fix build failure caused by wrong variable name in
 `DEBUG_KEXDH` section

---
 sshconnect2.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/sshconnect2.c b/sshconnect2.c
index 10e4f0a0..83bdf4f1 100644
--- a/sshconnect2.c
+++ b/sshconnect2.c
@@ -232,10 +232,10 @@ ssh_kex2(char *host, struct sockaddr *hostaddr, u_short port)
 
 #ifdef DEBUG_KEXDH
 	/* send 1st encrypted/maced/compressed message */
-	if ((r = sshpkt_start(ssh, SSH2_MSG_IGNORE)) != 0 ||
-	    (r = sshpkt_put_cstring(ssh, "markus")) != 0 ||
-	    (r = sshpkt_send(ssh)) != 0 ||
-	    (r = ssh_packet_write_wait(ssh)) != 0)
+	if ((r = sshpkt_start(active_state, SSH2_MSG_IGNORE)) != 0 ||
+	    (r = sshpkt_put_cstring(active_state, "markus")) != 0 ||
+	    (r = sshpkt_send(active_state)) != 0 ||
+	    (r = ssh_packet_write_wait(active_state)) != 0)
 		fatal("%s: %s", __func__, ssh_err(r));
 #endif
 }
-- 
2.37.1

